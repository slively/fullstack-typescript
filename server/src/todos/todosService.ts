import {todosDao} from 'todos/todosDao';
import {validateOrReject} from 'class-validator';
import {registerSchema} from 'class-validator';

registerSchema({
	name: 'todo',
	properties: {
		text: [{
			type: 'minLength',
			constraints: [1],
			options: undefined
		}]
	}
});

export const todosService = {
	find(): Promise<TodoEntity> {
		return todosDao.find();
	},

	create(todo: CreateTodoEntity): Promise<TodoEntity> {
		return validateOrReject('todo', todo)
			.then(() => todosDao.create(todo));
	},

	update(todo: Partial<TodoEntity>): Promise<TodoEntity> {
		return validateOrReject('todo', todo, { skipMissingProperties: true })
			.then(() => todosDao.update(todo));
	},

	deleteById(id: number): Promise<boolean> {
		return todosDao.deleteById(id);
	}
};
