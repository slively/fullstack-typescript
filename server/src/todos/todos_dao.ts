import { TimestampedPersistence } from 'lib/timestamped_persistence';
import { isSuccessfullyDeleted } from 'lib/dao_util';
import { getDatabaseClient } from 'database/db_client';
import * as Knex from 'knex';

export const TODOS_TABLE = 'todos';

export interface CreateTodoPersistence {
	text: string;
}

export interface TodoPersistence extends CreateTodoPersistence, TimestampedPersistence {
	id: number;
}

export class TodosDao {

	public static getInstance() {
		return this._instance || (this._instance = new TodosDao());
	}

	private static _instance: TodosDao;

	constructor(
		private dbClient: Knex = getDatabaseClient()
	) { }

	async find(): Promise<TodoPersistence> {
		return this.dbClient(TODOS_TABLE)
			.select('*')
			.orderBy('updatedAt', 'DESC');
	}

	async create(todo: CreateTodoPersistence): Promise<TodoPersistence> {
		return this.dbClient(TODOS_TABLE)
			.insert(todo)
			.returning('*')
			.get<TodoPersistence>(0);
	}

	async update(todo: Partial<TodoPersistence>): Promise<TodoPersistence> {
		return this.dbClient(TODOS_TABLE)
			.update({ ...todo, updatedAt: new Date(), createdAt: undefined })
			.where({ id: todo.id })
			.returning('*')
			.get<TodoPersistence>(0);
	}

	async deleteById(id: number): Promise<boolean> {
		return this.dbClient(TODOS_TABLE)
			.where({ id })
			.del()
			.then(isSuccessfullyDeleted);
	}
}