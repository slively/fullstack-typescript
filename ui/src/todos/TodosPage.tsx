import * as React from 'react';
import { TodosSearch } from 'todos/TodosSearch/TodosSearch';
import { TodosList } from 'todos/TodosList/TodosList';

export class TodosPage extends React.Component<{}, {}> {
	render() {
		return (
			<div>
				<h2>Todos</h2>
				<TodosSearch />
				<TodosList />
			</div>
		);
	}
}
