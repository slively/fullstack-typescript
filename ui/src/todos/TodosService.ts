import { CrudService } from 'lib/CrudService';
import { TodoEntity, CreateTodoEntity } from 'shared-models/todos';

export const todosService = new CrudService<TodoEntity, CreateTodoEntity>('/api/todos');
