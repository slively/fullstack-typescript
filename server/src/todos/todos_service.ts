import { TodosDao } from 'todos/todos_dao';
import { CreateTodoEntity, TodoEntity } from 'shared-models/todos';
import { validateOrReject } from 'class-validator';
import { registerSchema } from 'class-validator';

registerSchema({
	name: 'todo',
	properties: {
		text: [{
			type: 'minLength',
			constraints: [1],
			options: undefined
		}]
	}
});

export class TodosService {
	public static getInstance() {
		return this._instance || (this._instance = new TodosService());
	}

	private static _instance: TodosService;

	constructor(
		private todosDao: TodosDao = TodosDao.getInstance()
	) { }

	async find(): Promise<TodoEntity> {
		return this.todosDao.find();
	}

	async create(todo: CreateTodoEntity): Promise<TodoEntity> {
		await validateOrReject('todo', todo);
		return this.todosDao.create(todo);
	}

	async update(todo: Partial<TodoEntity>): Promise<TodoEntity> {
		await validateOrReject('todo', todo, { skipMissingProperties: true });
		return this.todosDao.update(todo);
	}

	async deleteById(id: number): Promise<boolean> {
		return this.todosDao.deleteById(id);
	}
}