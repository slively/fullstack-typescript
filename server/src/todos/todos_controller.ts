import { Request, Response } from 'express';
import { Router } from 'express';
import { TodosService } from 'todos/todos_service';

const todosService = TodosService.getInstance();

export const todosController =
	Router()
		.get('', (req: Request, res: Response) => res.sendPromise(todosService.find()))
		.post('', (req: Request, res: Response) => res.sendPromise(todosService.create(req.body)))
		.patch('', (req: Request, res: Response) => res.sendPromise(todosService.update(req.body)))
		.delete('/:id', (req: Request, res: Response) => res.sendPromise(todosService.deleteById(req.params.id)));
