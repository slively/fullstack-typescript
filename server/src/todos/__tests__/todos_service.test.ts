import { TodosService } from 'todos/todos_service';
import DoneCallback = jest.DoneCallback;
import { TodosDao } from '../todos_dao';

const todosDao = {
	find: jest.fn(),
	create: jest.fn(),
	update: jest.fn(),
	deleteById: jest.fn()
};
const todosDaoMock = new (jest.fn<TodosDao>(() => todosDao))();

const todosService = new TodosService(todosDaoMock);

describe('todosService', () => {
	beforeEach(() => jest.resetAllMocks());

	it('calls find on the dao', () => {
		todosService.find();

		expect(todosDao.find).toHaveBeenCalled();
	});

	it('calls create on the dao', () => {
		const todo = { text: 'foo', createdAt: new Date(), updatedAt: new Date() };

		return todosService.create(todo).then(
			() => expect(todosDao.create).toHaveBeenCalledWith(todo)
		);
	});

	it('throws a validation error when the new todo text has no length', (done: DoneCallback) => {
		const todo = { text: '' };

		todosService.create(todo).then(
			() => done.fail(),
			(err: any) => {
				expect(err).toMatchSnapshot();
				done();
			}
		);
	});

	it('calls update on the dao', () => {
		const todo = { text: 'foo' };

		return todosService.update(todo).then(
			() => expect(todosDao.update).toHaveBeenCalledWith(todo)
		);
	});

	it('calls deleteById on the dao', () => {
		const id = 101;

		todosService.deleteById(id);

		expect(todosDao.deleteById).toHaveBeenCalledWith(id);
	});
});
