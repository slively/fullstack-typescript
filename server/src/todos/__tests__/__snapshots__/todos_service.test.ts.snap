// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todosService throws a validation error when the new todo text has no length 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "minLength": "text must be longer than or equal to 1 characters",
    },
    "property": "text",
    "target": Object {
      "text": "",
    },
    "value": "",
  },
]
`;
